### PR TITLE
TensorFlow GPU

### PR DESCRIPTION
**Pull request by gppezzi**
@mschoengens

dependencies and patches are missing

### CODE REVIEW

File: easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+sanity_check_paths={
+    'files': ['bin/bazel'],
+    'dirs': ['bin'],
+}
+
+moduleclass = 'devel'
+
+
+# Build statistics
```
Comment (gppezzi) [inline easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb]: build logs/statistics shouldn't be included in this repository, please remove all buildstats from all files on this PR

File: easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Built with EasyBuild version 3.0.1 on 2017-01-02_15-19-02
```
Comment (gppezzi) [inline easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb]: please remove such "build with" comments.

File: easybuild/easyconfigs/p/protobuf-core/protobuf-core-3.0.2-CrayGNU-2016.11.eb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+easyblock = 'ConfigureMake'
+
+name = 'protobuf-core'
+version = '3.0.2'
```
Comment (gppezzi) [inline easybuild/easyconfigs/p/protobuf-core/protobuf-core-3.0.2-CrayGNU-2016.11.eb]: do you really need to keep two versions of protobuf on this PR?

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.11.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    ('cuDNN', '5.1', '', True),
+    ('SWIG', '3.0.10', '-Python-%(pyver)s'),
+    ('Bazel', '0.4.0'),
+    ('protobuf', '3.0.2', '-Python-%(pyver)s'),
+]
+
+sources = ['v%(version)s.tar.gz']
+source_urls = ['https://github.com/tensorflow/tensorflow/archive/']
+
+patches = [('/apps/daint/UES/6.0.UP02/sandbox-ds/soures/tensorflow-v%(version)s-cscs.patch')]
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.11.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb]: please don't use the absolute path for patches, you should include them on the PR and EB will search for them automatically.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-Python-3.5.2.eb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    ('Python','3.5.2'),
+    ('SWIG', '3.0.10', '-Python-%(pyver)s'),
+    ('Bazel', '0.4.0'),
+    ('protobuf', '3.1.0', '-Python-%(pyver)s'),
+]
+
+sources = ['v%(version)s.tar.gz']
+source_urls = ['https://github.com/tensorflow/tensorflow/archive/']
+
+patches = [('/apps/daint/UES/6.0.UP02/sandbox-ds/soures/tensorflow-v%(version)s-cscs.patch')]
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-Python-3.5.2.eb]: remove absolute path

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    ('cuDNN', '5.1', '', True),
+    ('SWIG', '3.0.10', '-Python-%(pyver)s'),
+    ('Bazel', '0.4.0'),
+    ('protobuf', '3.1.0', '-Python-%(pyver)s'),
+]
+
+sources = ['v%(version)s.tar.gz']
+source_urls = ['https://github.com/tensorflow/tensorflow/archive/']
+
+patches = [('/apps/daint/UES/6.0.UP02/sandbox-ds/soures/tensorflow-v%(version)s-gpu-cscs.patch')]
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb]: remove absolute path

File: easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb (hunk 2 above)
Comment (mschoengens) [inline easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb]: Done.

File: easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb (hunk 1 above)
Comment (mschoengens) [inline easybuild/easyconfigs/b/Bazel/Bazel-0.4.0-CrayGNU-2016.11.eb]: Done.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.11.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb (hunk 4 above)
Comment (mschoengens) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.11.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb]: Done.

File: easybuild/easyconfigs/s/SWIG/SWIG-3.0.10-CrayGNU-2016.11-Python-3.5.2.eb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+dependencies = [
+    ('Python', '3.5.2'),
+    ('PCRE', '8.38'),
+]
+
+configopts = 'LDFLAGS="$LDFLAGS -Wl,--rpath=$EBROOTPCRE/lib"'
+
+moduleclass = 'devel'
+
+# Build statistics
```
Comment (gppezzi) [inline easybuild/easyconfigs/s/SWIG/SWIG-3.0.10-CrayGNU-2016.11-Python-3.5.2.eb]: Please remove all build statistics

File: easybuild/easyconfigs/p/protobuf-core/protobuf-core-3.0.2-CrayGNU-2016.11.eb (hunk 3 above)
Comment (mschoengens) [inline easybuild/easyconfigs/p/protobuf-core/protobuf-core-3.0.2-CrayGNU-2016.11.eb]: 3.0.2 was required by TF 0.11

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-Python-3.5.2.eb (hunk 5 above)
Comment (mschoengens) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-Python-3.5.2.eb]: Done.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb (hunk 6 above)
Comment (mschoengens) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-0.12.0-CrayGNU-2016.11-cuda-8.0-Python-3.5.2.eb]: Done.

File: easybuild/easyconfigs/s/SWIG/SWIG-3.0.10-CrayGNU-2016.11-Python-3.5.2.eb (hunk 7 above)
Comment (mschoengens) [inline easybuild/easyconfigs/s/SWIG/SWIG-3.0.10-CrayGNU-2016.11-Python-3.5.2.eb]: Done.

File: easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.11.0-cuda-cscs.patch [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+   linker_flag: "-B/usr/bin/"
+ 
+ %{gcc_host_compiler_includes}
+-  tool_path { name: "gcov" path: "/usr/bin/gcov" }
++  tool_path { name: "gcov" path: "/opt/gcc/5.3.0/snos/bin/gcov" }
++
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/include"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include-fixed/"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include/"
++  cxx_builtin_include_directory: "/opt/nvidia/cudatoolkit8.0/default/include/"
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.11.0-cuda-cscs.patch]: how about $CRAY_CUDATOOLKIT_DIR/include ?

File: easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.11.0-cuda-cscs.patch [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+ CURRENT_DIR = os.path.dirname(sys.argv[0])
+-NVCC_PATH = CURRENT_DIR + '/../../../cuda/bin/nvcc'
+-LLVM_HOST_COMPILER_PATH = ('/usr/bin/gcc')
+-PREFIX_DIR = os.path.dirname(GCC_HOST_COMPILER_PATH)
++#NVCC_PATH = CURRENT_DIR + '/../../../cuda/bin/nvcc'
++#LLVM_HOST_COMPILER_PATH = ('/usr/bin/gcc')
++#PREFIX_DIR = os.path.dirname(GCC_HOST_COMPILER_PATH)
++
++
++NVCC_PATH = ('/opt/nvidia/cudatoolkit8.0/default/bin/nvcc')
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.11.0-cuda-cscs.patch]: $CRAY_CUDATOOLKIT_DIR/bin/nvcc

File: easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cuda-cscs.patch [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+ CURRENT_DIR = os.path.dirname(sys.argv[0])
+-NVCC_PATH = CURRENT_DIR + '/../../../cuda/bin/nvcc'
+-LLVM_HOST_COMPILER_PATH = ('/usr/bin/gcc')
+-PREFIX_DIR = os.path.dirname(GCC_HOST_COMPILER_PATH)
++#NVCC_PATH = CURRENT_DIR + '/../../../cuda/bin/nvcc'
++#LLVM_HOST_COMPILER_PATH = ('/usr/bin/gcc')
++#PREFIX_DIR = os.path.dirname(GCC_HOST_COMPILER_PATH)
++
++
++NVCC_PATH = ('/opt/nvidia/cudatoolkit8.0/default/bin/nvcc')
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cuda-cscs.patch]: $CRAY_CUDATOOLKIT_DIR/bin/nvcc

File: easybuild/easyconfigs/c/cuDNN/cuDNN-5.1.eb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+name = 'cuDNN'
+version = '5.1'
+
+homepage = 'https://developer.nvidia.com/cudnn'
+description = """Nvidias deep neural network library."""
+
+toolchain = {'name': 'dummy', 'version': 'dummy'}
+
+dependencies = [
+    ('cudatoolkit/8.0.44_GA_2.2.7_g4a6c213-2.1', EXTERNAL_MODULE),
```
Comment (gppezzi) [inline easybuild/easyconfigs/c/cuDNN/cuDNN-5.1.eb]: please update to the latest cuda patch version
```
   ('cudatoolkit/8.0.54_2.2.8_ga620558-2.1', EXTERNAL_MODULE),
```

File: easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cscs.patch [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+   linker_flag: "-B/usr/bin/"
+ 
+ %{gcc_host_compiler_includes}
+-  tool_path { name: "gcov" path: "/usr/bin/gcov" }
++  tool_path { name: "gcov" path: "/opt/gcc/5.3.0/snos/bin/gcov" }
++
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/include"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include-fixed/"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include/"
++  cxx_builtin_include_directory: "/opt/nvidia/cudatoolkit8.0/default/include/"
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cscs.patch]: $CRAY_CUDATOOLKIT_DIR/include

File: easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cuda-cscs.patch [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+   linker_flag: "-B/usr/bin/"
+ 
+ %{gcc_host_compiler_includes}
+-  tool_path { name: "gcov" path: "/usr/bin/gcov" }
++  tool_path { name: "gcov" path: "/opt/gcc/5.3.0/snos/bin/gcov" }
++
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/include"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include-fixed/"
++  cxx_builtin_include_directory: "/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/default/include/"
++  cxx_builtin_include_directory: "/opt/nvidia/cudatoolkit8.0/default/include/"
```
Comment (gppezzi) [inline easybuild/easyconfigs/t/TensorFlow/tensorflow-v0.12.0-cuda-cscs.patch]: $CRAY_CUDATOOLKIT_DIR/include

**Comment by gppezzi** [comment]
how about removing hardcoded paths for gcc:
```
sed -i s#/opt/gcc/5.3.0/#\$GCC_PATH/#g tensorflow-*.patch
```

**Comment by gppezzi** [comment]
@victorusu @mschoengens since we don't have a solution yet for the multicore version, I'm merging the GPU version. (MC version is now on the tf-mc branch).

The recipe can still be improved but at least this is a starting point.

**Pull request merged**